### ISSUE TITLE
ICE: failed to get layout for `E`: the type `E` has an unknown layout

### ISSUE THREAD

**Issue by petrochenkov** [rust-lang/rust#35988]
This is a regression: reproduces on beta and nightly, doesn't reproduce on stable.

``` rust
enum E {
    V([Box<E>])
}

fn main() {}
```

```
error: internal compiler error: ../src/librustc_lint/types.rs:698: failed to get layout for `E`: the type `E` has an unknown layout

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

thread 'rustc' panicked at 'Box<Any>', ../src/librustc_errors/lib.rs:634
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by eddyb** [comment]
This is not supported though, why aren't you getting an error? cc @arielb1 @nikomatsakis

**Comment by TimNN** [comment]
Note that referencing the type also causes an ICE on stable:

```rust
enum E {
    V([Box<E>])
}

fn foo(_: &E) {}

fn main() {}
```

This triggers without referencing the type since #34755 moved `variant_size_differences` from trans to a lint (I believe).

**Comment by jonas-schievink** [comment]
Yes, similar to #34571

**Comment by nikomatsakis** [comment]
This seems like a legitimate "staging" problem... not quite sure how to triage though.

**Comment by petrochenkov** [comment]
Fixed in 1.15.1, needs test.
```
rustc 1.15.1 (021bd294c 2017-02-08)
error[E0277]: the trait bound `[Box<E>]: std::marker::Sized` is not satisfied
 --> <anon>:2:7
  |
2 |     V([Box<E>])
  |       ^^^^^^^^^ the trait `std::marker::Sized` is not implemented for `[Box<E>]`
  |
  = note: `[Box<E>]` does not have a constant size known at compile-time
  = note: only the last field of a struct may have a dynamically sized type

error: aborting due to previous error
```
